api/src/main/java/io/onetable/spi/sync/TableFormatSync.java

```diff
@@ -21,46 +21,68 @@
 import java.time.Duration;
 import java.time.Instant;
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
 
-import lombok.AllArgsConstructor;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
 import lombok.extern.log4j.Log4j2;
 
 import io.onetable.model.IncrementalTableChanges;
 import io.onetable.model.OneSnapshot;
 import io.onetable.model.OneTable;
 import io.onetable.model.OneTableMetadata;
 import io.onetable.model.TableChange;
+import io.onetable.model.storage.TableFormat;
 import io.onetable.model.sync.SyncMode;
 import io.onetable.model.sync.SyncResult;
 
 /** Provides the functionality to sync from the OneTable format to the target format. */
 @Log4j2
-@AllArgsConstructor(staticName = "of")
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
 public class TableFormatSync {
-  private final TargetClient client;
+  private static final TableFormatSync INSTANCE = new TableFormatSync();
+
+  public static TableFormatSync getInstance() {
+    return INSTANCE;
+  }
 
   /**
    * Syncs the provided snapshot to the target table format.
    *
+   * @param targetClients the targets to sync with the snapshot
    * @param snapshot the snapshot to sync
    * @return the result of the sync process
    */
-  public SyncResult syncSnapshot(OneSnapshot snapshot) {
+  public Map<TableFormat, SyncResult> syncSnapshot(
+      Collection<TargetClient> targetClients, OneSnapshot snapshot) {
```

Comment:
I think it is a better here to take a single target client both from retrying per target client and also from a leaner API perspective.

Reply:
This blocks any sort of iterator approach to snapshot syncing in the future. Taking in all the clients allows us to operate over the source data once. In the future if we return an iterator of files in the snapshot to sync, we'd want to iterate over those files once

Reply:
I don't see why we still can't iterator approach with this change.

Reply:
You could use an iterator, but you will need to use an iterator per target instead of a single iterator. The idea is to allow this to be as efficient as possible, so only iterate over the source files once instead of per target client.

Reply:
Actually I'm not following the line of thought. I could understand what you are saying is related to incremental changes but not why it is related to Snapshot.

Reply:
In the future if we update the snapshot to also return an iterator of files instead of loading all the files into memory, we would use an iterator. That is not being implemented right now but I wanted to keep it consistent across the two sync modes.

Reply:
I see but I am not sure if we could get away with looking at one file at a time, because even in target we need to load current state for computing diffs and still be efficient.

Reply:
Since this depends on how we do things in future, we can leave in current state.